### ISSUE TITLE
Issue 003: `src` 非変更で Coarse experiment 用 train runner / criterion hook を追加する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#288]
# Issue 003: `src` 非変更で Coarse experiment 用 train runner / criterion hook を追加する

## 背景

site54 OOF の Coarse 学習は現在、以下の shell から通常 CLI を呼んでいる。

```text
proc/fbpick/site54/oof/scripts/run_coarse_train_fold.sh
  -> python cli/run_fbpick_coarse_train.py --config <01_coarse_train.yaml>
  -> seisai_engine.pipelines.fbpick.coarse.train.run_train()
```

`velocity_mask` や `trend_prior_loss` を experiment として学習時に適用するには、`packages/seisai-engine/src/**` を変更せずに、site54 OOF 側で train criterion / inner validation を差し替えられる hook が必要になる。

この issue では、runtime extension の土台だけを追加する。`velocity_mask` と `trend_prior_loss` の具体実装は後続 issue に分ける。

## 重要な制約

- `packages/seisai-engine/src/**` は変更しない。
- `packages/seisai-pick/src/**` は変更しない。
- default / baseline experiment では、既存 `cli/run_fbpick_coarse_train.py` と同じ結果になるように no-op 動作にする。
- fold list、train / inner_valid / heldout split は変更しない。
- `run_site54_oof_cv.py` の run-scoped layout を壊さない。

## 現在の repo 前提

関連ファイル:

```text
cli/run_fbpick_coarse_train.py
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/train.py
packages/seisai-engine/src/seisai_engine/pipelines/common/train_skeleton.py
packages/seisai-engine/src/seisai_engine/pipelines/common/train_skeleton_loop.py
packages/seisai-engine/src/seisai_engine/train_loop.py
proc/fbpick/site54/oof/scripts/run_coarse_train_fold.sh
proc/fbpick/site54/oof/scripts/make_coarse_fold_configs.py
```

`coarse.train.build_train_spec()` は `TrainSkeletonSpec` を返し、`criterion` と `infer_epoch_fn` が差し替え可能な形になっている。`TrainSkeletonSpec` は frozen ではない dataclass なので、`dataclasses.replace()` で新しい spec を作るのが安全。

## 目的

site54 OOF scripts 配下に、Coarse train spec を読み込み、必要な場合だけ criterion / infer_epoch_fn を wrap して `run_train_skeleton()` に渡す local runner を追加する。

## 変更対象

追加候補:

```text
proc/fbpick/site54/oof/scripts/run_fbpick_coarse_train_site54_experiment.py
proc/fbpick/site54/oof/scripts/site54_coarse_experiment_runtime.py
```

変更候補:

```text
proc/fbpick/site54/oof/scripts/run_coarse_train_fold.sh
proc/fbpick/site54/oof/scripts/make_coarse_fold_configs.py  # metadata section のみ必要なら
proc/fbpick/site54/oof/README.md                            # 必要最小限
```

変更禁止:

```text
packages/seisai-engine/src/**
packages/seisai-pick/src/**
```

## config interface

後続 issue が利用する top-level section を定義する。baseline ではこの section が無い、または `enabled: false` で no-op。

```yaml
site54_oof_experiment:
  coarse_train_runtime:
    enabled: true
    velocity_mask:
      enabled: false
    trend_prior_loss:
      enabled: false
    logging:
      write_runtime_summary: true
```

`enabled` が省略された場合は、`velocity_mask.enabled` または `trend_prior_loss.enabled` が true なら true と解釈してよい。ただし baseline 互換のため、section 自体が無い場合は完全 no-op。

### CLI

`run_fbpick_coarse_train_site54_experiment.py` は通常 CLI と同じ minimum interface にする。

```bash
python proc/fbpick/site54/oof/scripts/run_fbpick_coarse_train_site54_experiment.py \
  --config /workspace/proc/fbpick/site54/oof/runs/<run_id>/configs/fold00/01_coarse_train.yaml
```

任意で `--device` を追加してよいが、既存 CLI と挙動を揃えるため必須にしない。

### load / run

推奨実装:

```python
from dataclasses import replace
from pathlib import Path

from seisai_engine.pipelines.common import (
    load_cfg_with_base_dir,
    resolve_device,
    run_train_skeleton,
    seed_all,
)
from seisai_engine.pipelines.fbpick.coarse.train import build_train_spec

cfg, base_dir = load_cfg_with_base_dir(Path(config_path))
device = resolve_device(None)
spec = build_train_spec(cfg, base_dir=base_dir, device=device)
wrapped_spec = maybe_wrap_site54_coarse_train_spec(spec)
seed_all(wrapped_spec.seed_train)
run_train_skeleton(wrapped_spec)
```

`maybe_wrap_site54_coarse_train_spec(spec)` は `site54_oof_experiment.coarse_train_runtime` が無効なら `spec` をそのまま返す。

### criterion wrapper の形

後続 issue で velocity/trend を差し込めるよう、以下のような合成 criterion を用意する。

```python
class Site54CoarseExperimentCriterion:
    def __init__(self, base_criterion, runtime_cfg): ...

    def __call__(self, pred, target, batch):
        pred_for_loss = pred
        aux = {}
        # Issue 004: optional velocity_mask を pred_for_loss に適用
        # Issue 005: optional trend_prior_loss を追加
        base_loss = self.base_criterion(pred_for_loss, target, batch)
        total_loss = base_loss
        return total_loss
```

最初の実装では base criterion を呼ぶだけの no-op でよい。後続 issue が関数を追加しやすいように構造を整える。

### inner validation wrapper の形

`coarse.train._run_infer_epoch()` は `pred = model(x)` のあと criterion と debug PNG を実行している。src を変更しないため、必要なら local runner 側で同等の infer epoch 関数を持ち、runtime cfg に応じて `pred` を変換できるようにする。

この issue では no-op implementation でよい。後続 issue で `velocity_mask.apply_inner_valid` などを使う。

### batch tensor device helper

後続 issue で `offsets`, `dt_sec`, `fb_idx_coarse` を使うため、batch の一部だけを logits と同じ device に移す helper を用意しておく。

```python
def batch_with_tensor_keys_on_device(batch: dict, keys: list[str], device: torch.device) -> dict:
    out = dict(batch)
    for key in keys:
        if key in out and isinstance(out[key], torch.Tensor):
            out[key] = out[key].to(device=device, non_blocking=True)
    return out
```

元の `batch` は破壊しない。

## shell wiring

`run_coarse_train_fold.sh` は experiment runner を呼ぶように変更する。baseline no-op を守るなら、常に site54 runner を使ってもよい。

候補:

```bash
COARSE_TRAIN_ENTRYPOINT=${COARSE_TRAIN_ENTRYPOINT:-site54_experiment}
if [[ "$COARSE_TRAIN_ENTRYPOINT" == "default" ]]; then
  python cli/run_fbpick_coarse_train.py --config "$CONFIG"
else
  python proc/fbpick/site54/oof/scripts/run_fbpick_coarse_train_site54_experiment.py --config "$CONFIG"
fi
```

現行 script では `CONFIG_PATH` env を受け取っているように見えるが、実際には内部で `CONFIG="$CONFIG_ROOT/$FOLD/..."` を組み立てている。必要なら `CONFIG_PATH` が与えられた場合に優先する修正を入れてよい。ただし既存呼び出しは壊さない。

## runtime summary

extension が有効な場合だけ、`paths.out_dir` 配下に軽量 JSON を書くと比較しやすい。

```text
runs/<run_id>/experiments/<exp>/foldXX/01_coarse_train/site54_experiment_runtime.json
```

最初は以下程度でよい。

```json
{
  "enabled": true,
  "velocity_mask_enabled": false,
  "trend_prior_loss_enabled": false
}
```

baseline no-op では書かない、または `enabled: false` として書く。どちらにするかは README / tests と一致させる。

## 受け入れ条件

- `packages/seisai-engine/src/**` と `packages/seisai-pick/src/**` に変更がない。
- 新しい site54 train runner が `--config` で起動できる。
- `site54_oof_experiment.coarse_train_runtime` が無い config では base criterion / infer epoch のみを使う。
- `run_coarse_train_fold.sh` から新 runner を呼べる。
- `COARSE_TRAIN_ENTRYPOINT=default` などの escape hatch を用意した場合、従来 CLI も呼べる。
- py_compile が通る。
- smoke config で少なくとも dry-run / import smoke が通る。

## 確認コマンド

```bash
cd /workspace

python -m py_compile \
  proc/fbpick/site54/oof/scripts/run_fbpick_coarse_train_site54_experiment.py \
  proc/fbpick/site54/oof/scripts/site54_coarse_experiment_runtime.py

bash -n proc/fbpick/site54/oof/scripts/run_coarse_train_fold.sh

# config 生成済み前提の import / smoke 確認例
python proc/fbpick/site54/oof/scripts/run_site54_oof_cv.py \
  --run-id site54_runner_probe \
  --experiment baseline \
  --stage prepare_configs

python proc/fbpick/site54/oof/scripts/run_site54_oof_cv.py \
  --run-id site54_runner_probe \
  --experiment baseline \
  --stage coarse_train \
  --fold fold00 \
  --gpu cpu \
  --smoke
```
